alf/algorithms/one_step_loss.py

```diff
@@ -41,3 +44,139 @@ def __init__(self,
+            debug_summaries=debug_summaries,
+            name=name)
+
+        self._cdf_midpoints = (torch.arange(
+            num_quantiles, dtype=torch.float32) + 0.5) / num_quantiles
+        self._sum_over_quantiles = sum_over_quantiles
+
+    def compute_td_target(self, info, target_value):
+        """Calculate the td target.
+
+        The first dimension of all the tensors is time dimension and the second
+        dimesion is the batch dimension.
+
+        Args:
+            info (namedtuple): experience collected from ``unroll()`` or
+                a replay buffer. All tensors are time-major. ``info`` should
+                contain the following fields:
+                - reward:
+                - step_type:
+                - discount:
+            target_value (torch.Tensor): the time-major tensor for the value at
+                each time step. This is used to calculate return. ``target_value``
+                can be same as ``value``.
+        Returns:
+            td_target
+        """
+        if info.reward.ndim == 3:
+            # [T, B, 1]
+            assert info.reward.shape[-1] == 1, (
+                "Cannot handle multi-dimensional reward for OneStepTDQRLoss.")
+
+        value_shape = target_value.shape
+        rewards = info.reward.unsqueeze(-1).expand(value_shape)
+        step_types = info.step_type.unsqueeze(-1).expand(value_shape)
+        discounts = info.discount.unsqueeze(-1).expand(
+            value_shape) * self._gamma
+
+        returns = value_ops.one_step_discounted_return(
+            rewards=rewards,
+            values=target_value,
+            step_types=step_types,
+            discounts=discounts)
+
+        return returns
+
+    def forward(self, info, value, target_value):
+        """Calculate the loss.
+
+        The first dimension of all the tensors is time dimension and the second
+        dimesion is the batch dimension.
+
+        Args:
+            info (namedtuple): experience collected from ``unroll()`` or
+                a replay buffer. All tensors are time-major. ``info`` should
+                contain the following fields:
+                - reward:
+                - step_type:
+                - discount:
+            value (torch.Tensor): the time-major tensor for the value at each time
+                step. The loss is between this and the calculated return.
+            target_value (torch.Tensor): the time-major tensor for the value at
+                each time step. This is used to calculate return. ``target_value``
+                can be same as ``value``.
+        Returns:
+            LossInfo: with the ``extra`` field same as ``loss``.
+        """
+        assert value.ndim == 3 and target_value.ndim == 3, (
+            "input value and target_value should have shape (T, B, N)")
+        returns = self.compute_td_target(info, target_value)
+        value = value[:-1]
+
+        cdf_midpoints = self._cdf_midpoints.view(1, 1, 1, -1)
```

Comment:
It would be better to comment the shape of the tensor to help reading the code.

Reply:
```
# For QR-DQN, quantiles have a shape
# (T-1, B, n_quantiles). Thus, make self._cdf_midpoints
# broadcastable to (T-1, B, n_quantiles, n_quantiles)
```

Reply:
Added.